### DISCUSSION TITLE
Drag and Drop Functionality from Tree to Diagram in Sirius

### DISCUSSION THREAD

**Discussion by hoajie** [eclipse-sirius/sirius-desktop#488] [Q&A]
Hi Sirius team,

I hope you are doing well. I am currently working on a project using Sirius, where I have created a Tree Description. In this tree, I am using Diagram Creation to allow users to directly create diagrams from within the tree.

As shown in the image below, on the left side, there is a tree, and on the right side, there is a diagram that was created from the tree.

What I am trying to achieve:
I would like to implement a feature that allows dragging elements from the tree and dropping them directly onto the diagram on the right side. For example, in the attached image, I want to be able to drag an element like "action2" from the tree and drop it onto the diagram so that it is displayed there.

Question:
Could you provide any guidance or reference material on how to implement drag-and-drop functionality from a tree representation to a diagram in Sirius? Any advice or examples on how to get started with this feature would be greatly appreciated.

Thank you in advance for your help, and I look forward to your response!

Best regards,
![Snipaste_2024-10-24_17-15-49](https://github.com/user-attachments/assets/b4c8eddb-14cd-4671-bb23-4ba3ec742eb5)

**Comment by lredor** [comment]
Hi,
The test [DragAndDropFromTableAndTreeToDiagramTest](https://github.com/eclipse-sirius/sirius-desktop/blob/master/plugins/org.eclipse.sirius.tests.swtbot/src/org/eclipse/sirius/tests/swtbot/DragAndDropFromTableAndTreeToDiagramTest.java) seems to be a good starting point. The associated data is [here](https://github.com/eclipse-sirius/sirius-desktop/tree/master/plugins/org.eclipse.sirius.tests.swtbot/data/unit/dragAndDrop/tree_table_to_diagram).
Regards,
Laurent

**Reply by hoajie** [reply]
Hi Laurent,

Thank you so much for your prompt response and for pointing me to the DragAndDropFromTableAndTreeToDiagramTest. I really appreciate it!

I will definitely look into the test and the associated data you provided and will give it a try. Hopefully, this will help me achieve the drag-and-drop functionality I'm aiming for.

Thanks again for your guidance, and I’ll reach out if I need further assistance.

Best regards,
hoajie

**Reply by HuiHMegumin** [reply]
Hi, i'm wondering how to create unlimited reference on the diagram using drag n drop, is there any examples about it? @lredor

**Reply by lredor** [reply]
What do you mean by `unlimited reference`? The created elements are usually linked to the drag'n'dropped elements, therefore "limited" by the content of the selection.

**Reply by HuiHMegumin** [reply]
I mean create multiple references( reference the same instance ) on the diagram, I use create view to create it , but it will be deleted by refresh operation...
I debuged the code, it seems drag n drop and create view all works , because if i use getRepresentElement() method to get the list , it actual added , but after that , the code refreshs the diagram, then these duplicated reference has been deleted.
Thanks for reply.

**Reply by lredor** [reply]
It's normal. It works as designed. In Sirius, you can have only one graphical representation for a semantic element using a specific mapping.

**Reply by HuiHMegumin** [reply]
Yeah, because I tried all configs in odesign and it seems all cant create multiple reference on the diagram.
But if it's like what you said, the **Create View** feature is for what? It seems to create just the view(reference) that target is the corresponding instance.
And the strange thing is ,I've create a diagram that can create multiple reference on the diagram by accident, and if i use the same odesign create another diagram , it can't again...
Very thanks for you reply~

**Reply by lredor** [reply]
`Create View` can be useful when a diagram is not synchronized: In this mode, the graphical instances of the mapping are only created and deleted when explicitly requested by a tool.

**Reply by HuiHMegumin** [reply]
Yeah, so that means if i set the diagram as not synchronized,then i add the `create view` operation after drag n drop in **odesign**. After that i drag the instance from tree view to diagram, there will have a view created on the diagram. Is this description right?

**Reply by HuiHMegumin** [reply]
![image](https://github.com/user-attachments/assets/64c7f44a-8cba-46b2-a8c0-efbec5ea860f)
![image](https://github.com/user-attachments/assets/ea44c7a5-df4a-4542-9344-1951e6873f6e)

**Reply by HuiHMegumin** [reply]
![image](https://github.com/user-attachments/assets/4dc228e6-d9b8-4d43-a8fb-7cdc4e31930d)

**Reply by lredor** [reply]
`Is this description right?` Yes

**Reply by HuiHMegumin** [reply]
So the view can create only single one too? If i don't  add **Create View** operation, it still work, It's because dnd already add one , so the one that **Create View** created is duplicated and deleted by refresh.
